library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -478,48 +420,96 @@ impl<T, A: Allocator> VecDeque<T, A> {
         });
     }
 
+    /// Writes all values from `iter` to `dst`, wrapping
+    /// at the end of the buffer and returns the number
+    /// of written values.
+    ///
+    /// # Safety
+    ///
+    /// Assumes that `iter` yields at most `len` items.
+    /// Assumes capacity is sufficient.
+    unsafe fn write_iter_wrapping(
+        &mut self,
+        dst: usize,
+        mut iter: impl Iterator<Item = T>,
+        len: usize,
+    ) -> usize {
+        struct Guard<'a, T, A: Allocator> {
+            deque: &'a mut VecDeque<T, A>,
+            written: usize,
+        }
+
+        impl<'a, T, A: Allocator> Drop for Guard<'a, T, A> {
+            fn drop(&mut self) {
+                self.deque.len += self.written;
```

Comment:
I would suggest just having `len` in the guard, so that the `drop` doesn't do addition.

That way it'd match https://github.com/rust-lang/rust/blob/master/library/alloc/src/vec/set_len_on_drop.rs -- you might actually consider promoting that to `pub(crate)` in `alloc`, so `Vec` can keep using it and you can use it here too.

Reply:
I don't think it's really possible to just reuse the `SetLenOnDrop`, as it requires access to `&mut self.len` and the function simultaneously needs access to `&mut self` to write the elements. Also, I don't see why it's a big advantage to save the length as opposed to the number of written items; the addition in the `Drop` impl should be basically free, and overflow can't occur because the deque would panic due to capacity overflow before that could happen.

Reply:
Ah, I see, it's using method calls on self, so can't do the borrow splitting that the `Vec` one does.